File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int n = nums.size();
        if(n == 1) return (nums[0] == target); 
        if(nums[0] == target || nums[n-1] == target) return true;
        int l=1, r=n-2;
        while(l <= r){
            // while(l <= r && nums[l+1] == nums[l]) l++;
            // while(r >= l && nums[r-1] == nums[r]) r--;
            // if(l > r) return (nums[l] == target);
            int mid = (l+r)/2;
            if(nums[mid] == target) return true;
            if((nums[l] == nums[mid]) && (nums[r] == nums[mid])) {
                l++;
                r--;
            }
            else if(nums[mid] >= nums[l]){
                if((target >= nums[l]) && (target < nums[mid])) r = mid-1; 
                else l = mid+1;
            } else {
                if((target > nums[mid]) && (target <= nums[r])) l=mid+1; 
                else r = mid-1;
            }
        }
        return false;
    }
};

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums) - 1
        n = len(nums)

        while l <= r:
            
            while l < r and nums[l+1] == nums[l]:
                l += 1
            
            while r > l and nums[r] == nums[r-1]:
                r -= 1

            m = (l + r) // 2
            if nums[m] == target:
                return True
            elif nums[0] <= nums[m]:
                if nums[0] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if nums[m] < target <= nums[n - 1]:
                    l = m + 1
                else:
                    r = m - 1

        return False